t/04-openresty/ffi/304-proxy_wasm_set_property_getter.t

```diff
@@ -0,0 +1,597 @@
+]
+--- no_error_log
+[error]
+
+
+
+=== TEST 5: set_property_getter() - setting host properties getter works on body_filter_by_lua
+--- wasm_modules: hostcalls
+--- http_config
+    init_worker_by_lua_block {
+        local proxy_wasm = require "resty.wasmx.proxy_wasm"
+        local filters = {
+            { name = "hostcalls" },
+        }
+
+        _G.c_plan = assert(proxy_wasm.new(filters))
+        assert(proxy_wasm.load(_G.c_plan))
+    }
+--- config
+    location /t {
+        proxy_wasm hostcalls 'on=response_body \
+                              test=/t/log/property \
+                              name=wasmx.my_property';
+
+        rewrite_by_lua_block {
+            local proxy_wasm = require "resty.wasmx.proxy_wasm"
+            assert(proxy_wasm.attach(_G.c_plan))
+
+            assert(proxy_wasm.start())
+            ngx.say("ok")
+        }
+
+        body_filter_by_lua_block {
+            local proxy_wasm = require "resty.wasmx.proxy_wasm"
+
+            local function getter(key)
+                if key == "wasmx.my_property" then
+                    return true, "my_value"
+                end
+                return false
+            end
+
+            assert(proxy_wasm.set_property_getter(getter))
+        }
+    }
+--- response_body
+ok
+--- error_log eval
+[
+    qr/\[info\] .*? \[hostcalls\] on_response_body/,
+    qr/\[info\] .*? wasmx.my_property: my_value/,
+]
+--- no_error_log
+[error]
+
+
+
+=== TEST 6: set_property_getter() - setting host properties getter works on log_by_lua
+--- wasm_modules: hostcalls
+--- http_config
+    init_worker_by_lua_block {
+        local proxy_wasm = require "resty.wasmx.proxy_wasm"
+        local filters = {
+            { name = "hostcalls" },
+        }
+
+        _G.c_plan = assert(proxy_wasm.new(filters))
+        assert(proxy_wasm.load(_G.c_plan))
+    }
+--- config
+    location /t {
+        proxy_wasm_request_headers_in_access on;
+
+        proxy_wasm hostcalls 'on=log \
+                              test=/t/log/property \
+                              name=wasmx.my_property';
+
+        access_by_lua_block {
+            local proxy_wasm = require "resty.wasmx.proxy_wasm"
+            assert(proxy_wasm.attach(_G.c_plan))
+
+            assert(proxy_wasm.start())
+            ngx.say("ok")
+        }
+
+        log_by_lua_block {
+            local proxy_wasm = require "resty.wasmx.proxy_wasm"
+
+            local function getter(key)
+                if key == "wasmx.my_property" then
+                    return true, "my_value"
+                end
+                return false
+            end
+
+            assert(proxy_wasm.set_property_getter(getter))
+        }
+    }
+--- response_body
+ok
+--- error_log eval
+[
+    qr/\[info\] .*? \[hostcalls\] on_log/,
+    qr/\[info\] .*? wasmx.my_property: my_value/,
+]
+--- no_error_log
+[error]
+
+
+
+=== TEST 7: set_property_getter() - setting property getter at startup doesn't reset filter list
```

Comment:
It is not clear to me what this test is supposed to test; the startup seems to be the same as other tests, in fact this test seems functionally identical to `TEST 3 - sanity in access_by_lua`, except for `request_headers_in_access` flag which should not make any difference whatsoever? I don't see why this flag would be worth testing with host properties, but maybe am missing something. In my merge PR I marked these tests for getter/setter as `SKIP?` for now, given I'd rather get rid of them if they are redundant.

Reply:
@hishamhm I think this is my last item before merging; was there something I missed in this test that isn't present in any other test?

Reply:
You're right, tests 3 and 7 in this file are currently redundant, but the reason why this one originally existed is relevant and possibly should be captured in the test suite, if not with a test of its own, at least with a comment:

I based that test off test 4 of `proxy_wasm_set_property_host.t`, which is a regression test added when fixing a bug reported by the Gateway, which was an issue when setting properties between `attach` and `start`. So I replicated that test to ensure that configuring the property getter/setter also works when done in between `attach` and `start`. However, test 3 which is checking the `access_by_lua` phase is also doing that, which rendered the test case redundant.

So in short, we should test that the various phases, and we should test that setting the getter/setter works before and after `proxy_wasm.start`. So I'd say we have two options:

* test 3 should be changed so that the getter is set _after_ `proxy_wasm.start` in the access phase, and then test 7 is renamed to "setting property getter before proxy_wasm.start works"
* test 7 is removed, we assume that if setting the getter works in later phases then it also works on access after `proxy_wasm.start`, we keep test 3 as is, and then I'd recommend adding an explicit comment in the Lua block saying something like `-- ensuring that setting the getter in between attach and start works`

Reply:
Great thanks! Ok so I removed the test since the `header_filter_by_lua` test case does test this (set_prop after attach/start) while all other tests do attach/set_prop/start). I think this will be merged today.